hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -978,12 +978,15 @@ public void createDirectory(final Path path, final FileSystem.Statistics statist
           throws IOException {
     try (AbfsPerfInfo perfInfo = startTracking("createDirectory", "createPath")) {
       if (getAbfsConfiguration().getPrefixMode() == PrefixMode.BLOB) {
-        checkParentChainForFile(path, tracingContext);
+        ArrayList<Path> keysToCreateAsFolder = new ArrayList<>();
+        checkParentChainForFile(path, tracingContext, keysToCreateAsFolder);
 
         HashMap<String, String> metadata = new HashMap<>();
         metadata.put(X_MS_META_HDI_ISFOLDER, TRUE);
-        createFile(path, statistics, true,
-                permission, umask, tracingContext, metadata);
+        for (Path pathToCreate: keysToCreateAsFolder) {
+          createFile(pathToCreate, statistics, false,
```

Comment:
We should have under:
```
try {
createFile(pathToCreate, statistics, false,..)
} catch(FileAlreadyExistsException ex){}
```

Why???

Lets say there are two threads:
t1:fs.create("/d1/file1");
t2:fs.create("/d1/file2"):
```
Now lets say t1 check d1 and found its not there, prepare to create it....; context switch happens...
t2 check d1 not there, prepare to create and create.... context switch happens
t1 tries to create d1 -> FileAlreadyExistException .... the whole create fails for t1.